faiss/impl/NSG.cpp

```diff
@@ -0,0 +1,556 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+// -*- c++ -*-
+
+#include <faiss/impl/NSG.h>
+
+#include <algorithm>
+#include <mutex>
+#include <stack>
+#include <string>
+
+#include <faiss/impl/AuxIndexStructures.h>
+
+namespace faiss {
+
+namespace nsg {
+
+DistanceComputer *storage_distance_computer(const Index *storage) {
+  if (storage->metric_type == METRIC_INNER_PRODUCT) {
+    return new NegativeDistanceComputer(storage->get_distance_computer());
+  } else {
+    return storage->get_distance_computer();
+  }
+}
+
+}  // namespace faiss::nsg
+
+using namespace nsg;
+
+using LockGuard = std::lock_guard<std::mutex>;
+
+struct Neighbor {
+  int id;
+  float distance;
+  bool flag;
+
+  Neighbor() = default;
+  Neighbor(int id, float distance, bool f)
+      : id(id), distance(distance), flag(f) {}
+
+  inline bool operator<(const Neighbor &other) const {
+    return distance < other.distance;
+  }
+};
+
+struct SimpleNeighbor {
+  int id;
+  float distance;
+
+  SimpleNeighbor() = default;
+  SimpleNeighbor(int id, float distance) : id(id), distance(distance) {}
+
+  inline bool operator<(const SimpleNeighbor &other) const {
+    return distance < other.distance;
+  }
+};
+
+inline int insert_into_pool(Neighbor *addr, int K, Neighbor nn) {
+  // find the location to insert
+  int left = 0, right = K - 1;
+  if (addr[left].distance > nn.distance) {
+    memmove((char *)&addr[left + 1], &addr[left], K * sizeof(Neighbor));
+    addr[left] = nn;
+    return left;
+  }
+  if (addr[right].distance < nn.distance) {
+    addr[K] = nn;
+    return K;
+  }
+  while (left < right - 1) {
+    int mid = (left + right) / 2;
+    if (addr[mid].distance > nn.distance)
+      right = mid;
```

Comment:
I get a lot of warnings from the internal linter because if / while / for blocks should systematically be enclosed in {} 
Would you mind adding these in your code?

Reply:
OK, I will solve this.